2015/9-easy-exceptions-and-logging.pod

```diff
@@ -0,0 +1,293 @@
+
+Handling user errors can be a bit boring and tedious. We'll talk about a way
+to do so that requires very little code.
+
+First, let's have a look at how we might normally do it. I won't go into
+this in any detail, but you will probably recognise the concept. Maybe
+something like this:
+
+ package MyApp;
+
+ post '/settings' => sub {
+   my $message;
+   try {
+     MyApp::Settings->update($settings);
+     $message = "Settings updated successfully";
+   }
+   catch {
+     $message = $@;
+   };
+ };
+ 
+ package MyApp::Settings;
+
+ sub update {
+   ...
+   die "Invalid email address" unless ...
+ }
+
+That works fine to a degree, but the code can easily get more verbose than that,
+and even that alone can get a little tedious to keep typing.
+
+How would we do that with L<Dancer2::Plugin::LogReport>?
+
+ package MyApp;
+
+ post '/settings' => sub {
+   if (process sub{ MyApp::Settings::update($settings) }) {
+     success "Settings updated successfully";
+   }
+ };
+
+ package MyApp::Settings;
+
+ use Log::Report;
+ sub update {
+   ...
+   error "Invalid email address" unless ...
+ }
+
+That's it. But what if it wasn't successful? Well, the module handles all that
+for you. The C<error> function in the module is fatal just like C<die>, but as
+we showed above, its text will have automatically been sent to the template
+for display to the user.
+
+If you use C<error> outside of C<process>, then it is still fatal and the
+text is still sent to the template, but it's not handled quite as cleanly:
+the user is forwarded back to the same page if it was a C<POST> request,
+forwarded to C<'/'> if it was a C<GET> request, or fails uncleanly if it
+was already a C<GET> request to C<'/'>.
+
+As an aside, you would probably want to use the same route handler for both
+C<GET> and C<POST>. If the C<POST> is unsuccessful, the error is caught, and
+the page will behave as a C<GET> request (but with the error text rendered).
+
+=head3 A note on differentiating between expected and unexpected exceptions
+
+If a fatal error occurs that did not originate as a L<Log::Report>
+message, then at some point it has to be translated into its L<Log::Report>
+equivalent. There is a bit of guesswork that goes into this (see the
+L<Log::Report> documentation), and sometimes it might not quite be what
+you want. For example, if the code in the above example C<die>s due to
+a coding error such as calling a method on an undefined value, then
+it will be reported as an C<ERROR>, rather than C<PANIC>. The former
+will display the message to the user, which is not what we want. I'm
+considering changing the code to report all such errors at the C<PANIC> level, as they
+are normally always unexpected application exceptions.
+
+
+=head2 Dispatchers
+
+I have purposefully tried to keep this article simple, without going into
+too much of the terminology of L<Log::Report>. There is one thing I should
+mention though, and that is the concept of dispatchers. A dispatcher is
```

Comment:
I would rephrase this as:

"While you can use everything so far the way I have demonstrated, there is added value in understanding the concept of `Log::Report` dispatchers."